tests/interlocked_win32_ut/interlocked_win32_ut.c

```diff
@@ -0,0 +1,708 @@
+    ASSERT_ARE_EQUAL(uint64_t, 0xF0F0F0F0F0F0F0F0, return_val);
+}
+
+/*Tests_SRS_INTERLOCKED_WIN32_43_009: [interlocked_and_8 shall call InterlockedAnd8 from windows.h.]*/
+/*Tests_SRS_INTERLOCKED_WIN32_43_010 : [interlocked_and_8 shall return the initial value of * destination.]*/
+TEST_FUNCTION(interlocked_and_8_calls_InterlockedAdd8)
+{
+    ///arrange
+    volatile uint8_t destination = 0xF0;
+    uint8_t value = 0x0F;
+    STRICT_EXPECTED_CALL(mock_InterlockedAnd8((volatile char*)&destination, (char)value))
+        .SetReturn(0xF0);
+    ///act
+    uint8_t return_val = (uint8_t)interlocked_and_8((volatile int8_t*)&destination, (int8_t)value);
+
+    ///assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+    ASSERT_ARE_EQUAL(uint8_t, 0xF0, return_val);
+}
+
+/*Tests_SRS_INTERLOCKED_WIN32_43_011: [interlocked_compare_exchange shall call InterlockedCompareExchange from windows.h.]*/
+/*Tests_SRS_INTERLOCKED_WIN32_43_012 : [interlocked_compare_exchange shall return the initial value of * destination.]*/
+TEST_FUNCTION(interlocked_compare_exchange_calls_InterlockedCompareExchange)
+{
+    ///arrange
+    volatile int32_t destination = INT32_MAX;
+    int32_t comperand = INT32_MAX;
+    int32_t exchange = INT32_MIN;
+    STRICT_EXPECTED_CALL(mock_InterlockedCompareExchange((volatile LONG*)&destination, (LONG)exchange, (LONG)comperand))
+        .SetReturn(INT32_MAX);
+
+    ///act
+    int32_t return_val = interlocked_compare_exchange(&destination, exchange, comperand);
+
+    ///assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+    ASSERT_ARE_EQUAL(int32_t, INT32_MAX, return_val);
+}
+
+/*Tests_SRS_INTERLOCKED_WIN32_43_013: [interlocked_compare_exchange_128 shall call InterlockedCompareExchange128 from windows.h.]*/
+/*Tests_SRS_INTERLOCKED_WIN32_43_014 : [interlocked_compare_exchange_128 shall return true if* comperand_result equals the original value of* destination.]*/
+TEST_FUNCTION(interlocked_compare_exchange_128_calls_InterlockedCompareExchange128_true_case)
+{
+    ///arrange
+    volatile int64_t* destination = (volatile int64_t*)_aligned_malloc(2 * sizeof(int64_t), 16);
+    int64_t* comperand_result = (int64_t*)_aligned_malloc(2 * sizeof(int64_t), 16);
+
+    destination[0] = INT64_MAX;
+    destination[1] = INT64_MIN;
+    comperand_result[0] = INT64_MAX;
+    comperand_result[1] = INT64_MIN;
+
+    int64_t exchange_high = 2;
+    int64_t exchange_low = 3;
+
+    STRICT_EXPECTED_CALL(mock_InterlockedCompareExchange128((volatile LONG64*)destination, (LONG64)exchange_high, (LONG64)exchange_low, (LONG64*)comperand_result))
+        .SetReturn(1);
+    ///act
+    bool return_val = interlocked_compare_exchange_128(destination, exchange_high, exchange_low, comperand_result);
+
+    ///assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
```

Comment:
>ASSERT_ARE_EQUAL [](start = 4, length = 16)

ASSERT seems to be lacking comperand_result[0] == 3; [1]==2. Please add.

Reply:
I didn't add it in the unit test because that is checked in the integration test.

---
In reply to: [427744286](https://github.com/Azure/azure-c-util/pull/43#discussion_r427744286) [](ancestors = 427744286)